Add a `slimDOM` option to strip out unnecessary parts of the DOM

eoghanmurray opened the pull request:
unnecessary in terms of replay

 - <script> tags in the <head> take up unnecessary storage space and are often injected semi randomly to become a source of unnecessary variation between recordings of the same thing
 - comment tags can be stripped out without affecting display
 - future: this option could also turn on more aggressive stripping, e.g. elements that are hidden by CSS (assuming we can handle them becoming visible after mutation events)

eoghanmurray commented:
So open question on this that I don't know how to answer is, what happens when a mutation event happens on a script tag that is not in the mirror?  How could I handle that or is it okay if it generates an error?

Yuyz0112 commented:
I love the idea of aggressive stripping, so is it possible to rename the option like `strippingStrategy` with an object-based interface? Like something I've done in the sampling options:https://github.com/rrweb-io/rrweb/blob/master/src/types.ts#L134

BTW, since we already use a placeholder to display script contents: https://github.com/rrweb-io/rrweb-snapshot/blob/master/src/snapshot.ts#L303, is it still a big thing in storage size? (I agree with adding this to the strategy, just curious about it).

> So open question on this that I don't know how to answer is, what happens when a mutation event happens on a script tag that is not in the mirror? How could I handle that or is it okay if it generates an error?

I think we need to add some guard conditions in the mutation observer, such as ignore the events when the target's parent was not serialized.

eoghanmurray commented:
> is it possible to rename the option like strippingStrategy with an object-based interface
yes, I saw that and think it's a great idea for this too

> is it still a big thing in storage size? 
It's specific to how we are storing things.  Each time a different permutation of the <head> is generated, we have to store it separately on the backend. Looking further into how to normalize <head> to minimize this and I'll have the results of this change in a day or two to see how much of the variation is generated by non-deterministic injection of <script> tags (there could be non-deterministic injection of other things also).

> I think we need to add some guard conditions in the mutation observer
is this something you want me to do? I'm not sure what I should be looking for there

eoghanmurray commented:
> I think we need to add some guard conditions in the mutation observer

Yes this is definitely broken, I think mutations on absent nodes are getting an incorrect `parentId` assigned, which means we are attempting to .appendChild from e.g. text nodes (which is illegal)

eoghanmurray commented:
@Yuyz0112  I'm having trouble as I see that the logic for what should be stripped needs to be repeated in the mutation observer.  I see that this is also the case for `isBlocked`/`blockedClass`.  Is there a case to fold rrweb-snapshot back into rrweb or is there a better way to share this content?  

I was thinking of flagging an element as ignored using the `__sn` attribute, but that doesn't work for added nodes in a mutation.

eoghanmurray commented:
> Yes this is definitely broken, I think mutations on absent nodes are getting an incorrect parentId assigned
Ok, nevermind above comment, it was broken for me due to an unrelated change.

> I was thinking of flagging an element as ignored using the __sn attribute, but that doesn't work for added nodes in a mutation.
I've used this approach in https://github.com/rrweb-io/rrweb/pull/286 as I saw that nodes added via mutations also go through the `serializeNodeWithId` function.

This could all do with a review now!  Specifically:

 - is it okay that `serializeNodeWithId` returns null for the stripped out elements?
 - is it okay to flag these stripped elements with `.__sn.id == -2` ?

Yuyz0112 commented:
Sorry for the late @eoghanmurray 

The only thing I suggest is let rrweb-snasphot export the `-2` as constant then we can use it in rrweb.

eoghanmurray commented:
> The only thing I suggest is let rrweb-snasphot export the `-2` as constant then we can use it in rrweb.

Thanks! Have added that now and rebased on latest in this branch and the rrweb one.

Only thing todo now is to add an object based interface, but I think that could be added at a later date; maybe we could wait til someone doesn't like some aspect of the slimming, or if someone wants to add more aggressive stripping.  

I've added in the following (currently disabled due to anticipated affect on performance), but which might be a good example of an optional extra:

```
// would impede performance: 
getComputedStyle(n)['white-space'] === 'normal'
    )) {
      preserveWhiteSpace = false;
```

Other things I'd like to strip out are certain attributes, but there's always a possibility that they are used in some CSS rule, so unless we can cheaply inspect css rules I think it's a non-runner.  I wish there was a version of `getComputedStyle` that just listed the rules applied to an element, rather than calculating out the actual values (along with associated cost of laying out).

eoghanmurray commented:
@Yuyz0112 exactly what is stripped is now configurable via a `slimDOMOptions` (modelled on `maskInputOptions`).  At the moment there is very fine grained control over which meta tags should be removed — this could be consolidated if required.

Can you let me know if there's anything blocking this being merged?
With thanks!

Yuyz0112 commented:
Thanks, going to merge.

eoghanmurray commented:
Great, thanks!
I'm gonna do a test where I strip out nearly all `<meta>` tags from recordings across a range of websites, to see which ones (if any) actually are important in terms of playback.  Depending on how that plays out we could switch to a metatag whitelist rather than the current blacklist.
